File: Napakalaki/src/napakalaki/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 *
 * @author alexnp
 */
public class CardDealer {
    public static CardDealer instance = null;
    
    public static CardDealer getInstance() {
        if(instance == null) { instance = new CardDealer(); }
        
        return instance;
    }
    
     List<Treasure> unusedTreasures,
            usedTreasures;
    
     List<Monster> unusedMonsters,
            usedMonsters;
    
    private CardDealer() {
        usedTreasures = new ArrayList<>();
        unusedTreasures = new ArrayList<>();
        usedMonsters = new ArrayList<>();
        unusedMonsters = new ArrayList<>();
    }
    
    private void initTreasureCardDeck() {
        unusedTreasures.clear();
        
        unusedTreasures.add(
                new Treasure("¡Si, mi amo!", 4, TreasureKind.HELMET));

        unusedTreasures.add(
                new Treasure("Botas de investigación", 3, TreasureKind.SHOES));

        unusedTreasures.add(
                new Treasure("Capucha de Cthulhu", 3, TreasureKind.HELMET));

        unusedTreasures.add(
                new Treasure("A prueba de balas", 2, TreasureKind.ARMOR));

        unusedTreasures.add(
                new Treasure("Botas de lluvia ácida", 1, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Casco minero", 2, TreasureKind.HELMET));

        unusedTreasures.add(
                new Treasure("Ametralladora ACME", 4, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Camiseta de la ETSIIT", 1, TreasureKind.ARMOR));

        unusedTreasures.add(
                new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Fez alópodo", 3, TreasureKind.HELMET));

        unusedTreasures.add(
                new Treasure("Hacha prehistorica", 2, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR));

        unusedTreasures.add(
                new Treasure("Gaita", 4, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Insecticida", 2, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Escopeta de tres cañones", 4, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Garabato mistico", 2, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("La rebeca metalica", 2, TreasureKind.ARMOR));

        unusedTreasures.add(
                new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Necro-comicon", 1, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Necronomicon", 5, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Linterna a dos manos", 3, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Necro-gnomicon", 2, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Necrotelecom", 2, TreasureKind.HELMET));

        unusedTreasures.add(
                new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Necro-playboycon", 3, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Porra preternatural", 2, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS));

        unusedTreasures.add(
                new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND));

        unusedTreasures.add(
                new Treasure("Tentaculo de pega", 2, TreasureKind.HELMET));

        unusedTreasures.add(
                new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOES));


    }
    
    private void initMonsterCardDeck() {
        unusedMonsters.clear();
        
        unusedMonsters.add(
                new Monster("3 Byakhees de bonanza", 8,
                        new Prize(2, 1),
                        new BadConsequence(
                                "Pierdes tu armadura visible y otra oculta", 0,
                                Arrays.asList(TreasureKind.ARMOR),
                                Arrays.asList(TreasureKind.ARMOR))));
        unusedMonsters.add(
                new Monster("Tenochtitlan", 2,
                        new Prize(1, 1),
                        new BadConsequence("Embobados con el lindo primigenio te descartas de tu casco visible", 0,
                                Arrays.asList(TreasureKind.HELMET),
                                Collections.<TreasureKind>emptyList())));
        unusedMonsters.add(
                new Monster("El sopor de Dunwich", 2,
                        new Prize(1, 1),
                        new BadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible", 0,
                                Arrays.asList(TreasureKind.SHOES),
                                Collections.<TreasureKind>emptyList())));
        unusedMonsters.add(
                new Monster("Demonios de Magaluf", 2,
                        new Prize(4, 1),
                        new BadConsequence("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta", 0, 1, 1)));
        unusedMonsters.add(
                new Monster("El gorrón en el umbral", 13,
                        new Prize(3, 1),
                        new BadConsequence("Pierdes todos tus tesoros visibles", 0, Integer.MAX_VALUE, 0)));
        unusedMonsters.add(
                new Monster("H.P. Munchcraft", 6, 
                        new Prize(2, 1), 
                        new BadConsequence("Pierdes la armadura visible", 0,
                                Arrays.asList(TreasureKind.ARMOR), 
                                Collections.<TreasureKind>emptyList())));
        
        unusedMonsters.add(
                new Monster("Necrófago", 13, 
                        new Prize(1, 1), 
                        new BadConsequence("Sientes bichos bajo la ropa. Descarta la armadura visible", 0, 
                                Arrays.asList(TreasureKind.ARMOR), Collections.<TreasureKind>emptyList())));
        
        unusedMonsters.add(
                new Monster("El rey de rosado", 11, 
                        new Prize(3, 2), 
                        new BadConsequence("Pierdes 5 niveles de 3 tesoros visibles", 5, 3, 0)));
        
        unusedMonsters.add(
                new Monster("Flecher", 2, 
                        new Prize(1, 1), 
                        new BadConsequence("Toses los pulmones y pierdes 2 niveles", 2, 0, 0)));
        
        unusedMonsters.add(
                new Monster("Los hondos", 8, 
                        new Prize(2, 1), 
                        new BadConsequence("Estos monstruos resultan bastantes superficiales y te aburren mortalmente. Estas muerto", true)));
        
        unusedMonsters.add(
                new Monster("Semillas Cthulhu", 4, 
                        new Prize(2, 1), 
                        new BadConsequence("Pierdes 2 niveles y 2 tesoros ocultos.", 2, 0, 2)));
        
        unusedMonsters.add(
                new Monster("Dameargo", 1, 
                        new Prize(2, 1), 
                        new BadConsequence("Te intentas escaquear. Pierdes una mano visible.", 0,
                                Arrays.asList(TreasureKind.ONEHAND), 
                                Collections.<TreasureKind>emptyList())));
        unusedMonsters.add(
                new Monster("Pollipólipo volante", 3, 
                        new Prize(2, 1), 
                        new BadConsequence("Da mucho asquito. Pierdes 3 niveles", 3, 0, 0)));
        unusedMonsters.add(
                new Monster("Yskhtihyssg-Goth", 14, 
                        new Prize(3, 1), 
                        new BadConsequence("No le hace gracia que pronuncien mal su nombre. Estas muerto", true)));
        unusedMonsters.add(
                new Monster("Familia feliz", 1, 
                        new Prize(3, 1), 
                        new BadConsequence("La familia te atrapa. Estás muerto.", true)));
        unusedMonsters.add(
                new Monster("Roboggoth", 8, 
                        new Prize(2, 1), 
                        new BadConsequence("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible", 2, Arrays.asList(TreasureKind.BOTHHANDS), Collections.<TreasureKind>emptyList())));
        unusedMonsters.add(
                new Monster("El espía sordo", 5, 
                        new Prize(1, 1), 
                        new BadConsequence("Te asusta en la noche. Pierdes un casco visible", 0,
                                Arrays.asList(TreasureKind.HELMET), Collections.<TreasureKind>emptyList())));
        unusedMonsters.add(
                new Monster("Tongue", 19, 
                        new Prize(2, 1), 
                        new BadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles", 2, 5, 0)));
        unusedMonsters.add(
                new Monster("Bicéfalo", 21, 
                        new Prize(2, 1), 
                        new BadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos", 3, Integer.MAX_VALUE, 0)));
    
    }
    
    private void shuffleTreasures() {
            Collections.shuffle(unusedTreasures);
    }
    
    private void shuffleMonster() {
            Collections.shuffle(unusedMonsters);
    }
    
    public Treasure nextTreasure() {
            if(unusedTreasures.size() == 0) {
                shuffleTreasures();
                
                List<Treasure> l = usedTreasures;
                usedTreasures = unusedTreasures;
                unusedTreasures = l;
            }
            
            return unusedTreasures.remove(0);
    }
    
    public Monster nextMonster() {
            if(unusedMonsters.size() == 0) {
                shuffleTreasures();
                
                List<Monster> l = usedMonsters;
                usedMonsters = unusedMonsters;
                unusedMonsters = l;
            }
            
            return unusedMonsters.remove(0);
    }
    
    public void giveTreasureBack(Treasure t) {
            usedTreasures.add(t);
    }
    
    public void giveMonsterBack(Monster m) {
            usedMonsters.add(m);
    }
    
    public void initCards() {
            initMonsterCardDeck();
            initTreasureCardDeck();
    }
}

File: Napakalaki-rb/lib/CardDealer.rb
#encoding: utf-8

require 'singleton'

require 'Monster.rb'
require 'Treasure.rb'

class CardDealer
  include Singleton
  
  def initialize
    @used_monsters = []
    @used_treasures = []
    @unused_monsters = []
    @unused_treasures = []
  end
  
  attr_reader :unused_monsters, :unused_treasures
  

  def init_treasure_card_deck
    @unused_treasures = [
      Treasure.new("¡Si, mi amo!", 4, TreasureKind::HELMET),
      Treasure.new("Botas de investigación", 3, TreasureKind::SHOES),
      Treasure.new("Capucha de Cthulhu", 3, TreasureKind::HELMET),
      Treasure.new("A prueba de balas", 2, TreasureKind::ARMOR),
      Treasure.new("Botas de lluvia ácida", 1, TreasureKind::BOTHHANDS),
      Treasure.new("Casco minero", 2, TreasureKind::HELMET),
      Treasure.new("Ametralladora ACME", 4, TreasureKind::BOTHHANDS),
      Treasure.new("Camiseta de la ETSIIT", 1, TreasureKind::ARMOR),
      Treasure.new("Clavo de rail ferroviario", 3, TreasureKind::ONEHAND),
      Treasure.new("Cuchillo de sushi arcano", 2, TreasureKind::ONEHAND),
      Treasure.new("Fez alópodo", 3, TreasureKind::HELMET),
      Treasure.new("Hacha prehistorica", 2, TreasureKind::ONEHAND),
      Treasure.new("El aparato del Pr. Tesla", 4, TreasureKind::ARMOR),
      Treasure.new("Gaita", 4, TreasureKind::BOTHHANDS),
      Treasure.new("Insecticida", 2, TreasureKind::ONEHAND),
      Treasure.new("Escopeta de tres cañones", 4, TreasureKind::BOTHHANDS),
      Treasure.new("Garabato mistico", 2, TreasureKind::ONEHAND),
      Treasure.new("La rebeca metalica", 2, TreasureKind::ARMOR),
      Treasure.new("Lanzallamas", 4, TreasureKind::BOTHHANDS),
      Treasure.new("Necro-comicon", 1, TreasureKind::ONEHAND),
      Treasure.new("Necronomicon", 5, TreasureKind::BOTHHANDS),
      Treasure.new("Linterna a dos manos", 3, TreasureKind::BOTHHANDS),
      Treasure.new("Necro-gnomicon", 2, TreasureKind::ONEHAND),
      Treasure.new("Necrotelecom", 2, TreasureKind::HELMET),
      Treasure.new("Mazo de los antiguos", 3, TreasureKind::ONEHAND),
      Treasure.new("Necro-playboycon", 3, TreasureKind::ONEHAND),
      Treasure.new("Porra preternatural", 2, TreasureKind::ONEHAND),
      Treasure.new("Shogulador", 1, TreasureKind::BOTHHANDS),
      Treasure.new("Varita de atizamiento", 3, TreasureKind::ONEHAND),
      Treasure.new("Tentaculo de pega", 2, TreasureKind::HELMET),
      Treasure.new("Zapato deja-amigos", 1, TreasureKind::SHOES)
    ]
  end
  
  def init_monster_card_deck
    @unused_monsters = [
      Monster.new("3 Byakhees de bonanza", 8,
              Prize.new(2, 1), 
              BadConsequence.new_level_specific_treasures("Pierdes tu armadura visible y otra oculta", 0, 
                [TreasureKind::ARMOR], [TreasureKind::ARMOR])),

      Monster.new("Tenochtitlan", 2,
              Prize.new(1, 1),
              BadConsequence.new_level_specific_treasures("Embobados con el lindo primigenio te descartas de tu casco visible", 0,
                [TreasureKind::HELMET], [])),

      Monster.new("El sopor de Dunwich", 2,
              Prize.new(1, 1),
              BadConsequence.new_level_specific_treasures("El primordial bostezo contagioso. Pierdes el calzado visible", 0,
                      [TreasureKind::SHOES], [])),

      Monster.new("Demonios de Magaluf", 2,
              Prize.new(4, 1),
              BadConsequence.new_level_num_treasures("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta", 0, 1, 1)),

      Monster.new("El gorrón en el umbral", 13,
              Prize.new(3, 1),
              BadConsequence.new_level_num_treasures("Pierdes todos tus tesoros visibles", 0, 0, 0)),

      Monster.new("H.P. Munchcraft", 6, 
              Prize.new(2, 1), 
              BadConsequence.new_level_specific_treasures("Pierdes la armadura visible", 0,
                      [TreasureKind::ARMOR], [])),

      Monster.new("Necrófago", 13, 
              Prize.new(1, 1), 
              BadConsequence.new_level_specific_treasures("Sientes bichos bajo la ropa. Descarta la armadura visible", 0, 
                      [TreasureKind::ARMOR], [])),

      Monster.new("El rey de rosado", 11, 
              Prize.new(3, 2), 
              BadConsequence.new_level_num_treasures("Pierdes 5 niveles de 3 tesoros visibles", 5, 3, 0)),

      Monster.new("Flecher", 2, 
              Prize.new(1, 1), 
              BadConsequence.new_level_num_treasures("Toses los pulmones y pierdes 2 niveles", 2, 0, 0)),

      Monster.new("Los hondos", 8, 
              Prize.new(2, 1), 
              BadConsequence.new_death("Estos monstruos resultan bastantes superficiales y te aburren mortalmente. Estas muerto")),

      Monster.new("Semillas Cthulhu", 4, 
              Prize.new(2, 1), 
              BadConsequence.new_level_num_treasures("Pierdes 2 niveles y 2 tesoros ocultos.", 2, 0, 2)),

      Monster.new("Dameargo", 1, 
              Prize.new(2, 1), 
              BadConsequence.new_level_specific_treasures("Te intentas escaquear. Pierdes una mano visible.", 0,
                [TreasureKind::ONEHAND], [])),

      Monster.new("Pollipólipo volante", 3, 
              Prize.new(2, 1), 
              BadConsequence.new_level_num_treasures("Da mucho asquito. Pierdes 3 niveles", 3, 0, 0)),

      Monster.new("Yskhtihyssg-Goth", 14, 
              Prize.new(3, 1), 
              BadConsequence.new_death("No le hace gracia que pronuncien mal su nombre. Estas muerto")),

      Monster.new("Familia feliz", 1, 
              Prize.new(3, 1), 
              BadConsequence.new_death("La familia te atrapa. Estás muerto.")),

      Monster.new("Roboggoth", 8, 
              Prize.new(2, 1), 
              BadConsequence.new_level_specific_treasures("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible", 2, 
                [TreasureKind::BOTHHANDS], [])),

      Monster.new("El espía sordo", 5, 
              Prize.new(1, 1), 
              BadConsequence.new_level_specific_treasures("Te asusta en la noche", 0,
                [TreasureKind::HELMET], [])),

      Monster.new("Tongue", 19, 
              Prize.new(2, 1), 
              BadConsequence.new_level_num_treasures("Menudo susto te llevas", 2, 5, 0)),

      Monster.new("Bicéfalo", 21, 
              Prize.new(2, 1), 
              BadConsequence.new_level_num_treasures("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos", 3, 100000, 0)),
    ]
  end
  
  def shuffle_treasures
    @unused_treasures.shuffle!
  end
  
  def shuffle_monsters
    @unused_monsters.shuffle!
  end
 
public
  def give_treasure_back(t)
    @used_treasures.push(t)
  end
  
  def give_monster_back(m)
    @used_monsters.push(m)
  end
  
  def init_cards
    init_monster_card_deck
    init_treasure_card_deck
  end
  
  def next_monster
    
  end
  
  def next_treasure
    
  end
  
end
